R/utils-latency.R

```diff
@@ -0,0 +1,191 @@
+#' offset each relevant column by it's appropriate latency
+#' works for either adjusting aheads or lags
+#' @param shift_cols a tibble which must have the columns `column`, the name of
+#'   the column to adjust, `latency` the latency of the original column relative
+#'   to the `as_of` date, `new_name`, the names in `column` adjusted by the
+#'   latencies `latency`
+#' @param new_data just what is says
+#' @param keys the variables which are used as keys
+#' @keywords internal
+extend_either <- function(new_data, shift_cols, keys) {
+  shifted <-
+    shift_cols %>%
+    select(-any_of(c("shifts", "effective_shift", "type", "role", "source"))) %>%
+    pmap(\(original_name, latency, new_name) {
+      epi_shift_single(
+        x = new_data,
+        col = original_name,
+        shift_val = latency,
+        newname = new_name,
+        key_cols = keys
+      )
+    }) %>%
+    map(\(x) zoo::na.trim(x)) %>%
+    reduce(
+      dplyr::full_join,
+      by = keys
+    )
+
+  return(new_data %>%
+    select(-all_of(shift_cols$original_name)) %>% # drop the original versions
+    dplyr::full_join(shifted, by = keys) %>%
+    dplyr::group_by(dplyr::across(dplyr::all_of(keys[-1]))) %>%
+    dplyr::arrange(time_value) %>%
+    dplyr::ungroup())
+}
+
+#' find the columns added with the lags or aheads, and the amounts they have
+#' been changed
+#' @param prefix the prefix indicating if we are adjusting lags or aheads
+#' @param new_data the data transformed so far
+#' @return a tibble with columns `column` (relevant shifted names), `shift` (the
+#'   amount that one is shifted), `latency` (original columns difference between
+#'   max_time_value and as_of (on a per-initial column basis)),
+#'   `effective_shift` (shifts+latency), and `new_name` (adjusted names with the
+#'   effective_shift)
+#' @keywords internal
+#' @importFrom stringr str_match
+#' @importFrom dplyr rowwise %>%
+#' @importFrom magrittr %<>%
+get_shifted_column_tibble <- function(
+    prefix, new_data, terms_used, as_of, latency,
+    sign_shift, info, call = caller_env()) {
+  relevant_columns <- names(new_data)[grepl(prefix, names(new_data))]
+  to_keep <- rep(FALSE, length(relevant_columns))
+  for (col_name in terms_used) {
+    to_keep <- to_keep | grepl(col_name, relevant_columns)
+  }
+  relevant_columns <- relevant_columns[to_keep]
+  if (length(relevant_columns) == 0) {
+    cli::cli_abort("There is no column(s) {terms_used}.",
+      current_column_names = names(new_data),
+      class = "epipredict_adjust_latency_nonexistent_column_used",
+      call = call
+    )
+  }
+  # this pulls text that is any number of digits between two _, e.g. _3557_, and
+  # converts them to an integer
+  shift_amounts <- stringr::str_match(relevant_columns, "_(\\d+)_") %>%
+    `[`(, 2) %>%
+    as.integer()
```

Comment:
I'm a bit concerned about the robustness of this. For example, if I did `step_epi_lag(..., prefix = "la_g_",...)`, this would fail. I think we need to examine the recipe (then you can just capture the lags/leads directly).

Reply:
at the moment `step_adjust_latency` also has a `prefix` argument, so if you're setting it in `step_epi_lag` you'd also set it in `step_adjust_latency`, and would avoid the problem. I guess if you started choosing prefixes like `lag_3_` so that  there'd be multiple numbers it would break.

I'll look into grabbing the lags/leads at initial construction time again. I think there may be some ambiguity about which previous step to be inheriting from, but that's probably resolvable.

Reply:
right, the problem is that `recipes_eval_select`/ the actual `terms` isn't available during initial construction, so I don't actually know which columns the user is trying to get at. And `prep` doesn't know anything about the preceding functions.